packages/freezed/lib/src/freezed_generator.dart

```diff
@@ -186,6 +189,11 @@ Read here: https://github.com/rrousselGit/freezed/tree/master/packages/freezed#t
     return false;
   }
 
+  bool _hasGenericArgumentFactories(ClassElement element) {
```

Comment:
For union types, what if one constructor uses `hasGenericArgument` but another doesn't?

We probably should check this case and throw an error

Reply:
Similarly, we probably want a:

```dart
@Freezed(hasGenericArgument: true)
```

to avoid having to pass a `@JsonSerializable` on all the constructors of a union

Reply:
If the other constructor does not use the `hasGenericArgument` annotation and has a generic argument, json_serializable throws an error.

Reply:
it may not if we rely on TypeConverter, and we probably shouldn't rely on json_serializable for this anyway. Freezed should have its own error handling

Reply:
I am sorry, did you mean `TypeChecker`?, I am not familiar with `TypeConverter`

Reply:
TypeConverted is from JsonSerializable

We still need the `@Freezed(hasGenericArgument: true)` annotation and the exception

Reply:
Oh, I think you mean `JsonConverter`?

But still I'm failing to see what the use case could be, because something like this would not require argument factories for the generic types:

```dart
@Freezed(genericArgumentFactories: true)
class Generic<T> with _$Generic<T> {
  factory Generic(@DataConverter() T a) = _Generic<T>;

  factory Generic.fromJson(Map<String, dynamic> json) =>
      _$GenericFromJson<T>(json);
}

class DataConverter<T> implements JsonConverter<T, Object?> {
  const DataConverter();

  @override
  T fromJson(Object? json) {
    return json as T;
  }

  @override
  Object? toJson(T object) {
    return object;
  }
}
```

The only use case I can think of is when using the `generic_argument_factories` option for `json_serializable` in the `buld.yaml` file. In such case we can not rely in the `@JsonSerializable` annotation so it would be required to hint freezed using `@Freezed(genericArgumentFactories: true)` at the class level.

Reply:
It's not really about a use-case. It's about not having Freezed depend on how json_serializable behaves.

We need Freezed to provide a custom error message

Reply:
I've added an assertion to verify that if one constructor is annotated with `@JsonSerializable(genericArgumentFactories: true)` all other constructors are annotated as well.

https://github.com/fabianMendez/freezed/blob/9a9fb9d658af4f0892b709825ebff76236c99784/packages/freezed/lib/src/freezed_generator.dart#L158-L178

What do you think?